### ISSUE TITLE
Not a bug but a question about the emulator fork itself, there was no issues tab there.

### ISSUE THREAD

**Issue by radianthero** [PTwr/0079tl#17]
Wanted to say thank you for making a great tool, I'm currently using it for a translation project right now but have a question. It seems the default behavior of the subtitles is to fade out, is there a way to have it so it doesn't fade out? I'm using this tool to translate an FMV cutscene and its a bit weird to see the text disappear so suddenly. 

Thanks!

**Comment by AdamZielinski-Devtactics** [comment]
Hi, Dolphin subtitle system is a dirty hack abusing its notification system. Its just reusing whatever system was in place to display system messages. As such there are multiple shortcomings, like crappy pixel-font that looks horribly when text is larger.
I think I remember bits of code where feade-out behavior could be changed.

As you probably noticed, subtitle format is rather crappy and has to struggle against limitations of both Wii and rest of Dolphin.
It is all just an experimental implementation using dirty hack, and movie-subs are even more experimental. To put it bluntly, I had no idea what I was doing, what I needed, and how to glue it all together, and then kept crashing against all the weird quirks of emulation.

 It is unfinished, and highly experimental, but I plan to get back to it "soon". Rough roadmap would be:
- start from scratch on newest Dolphin sourcode
- port simplified text display thingie
- fix all the memory leaks in JSON parsing...
- improve file offset logger
- make MR with some nice examples of usage to get it into vanilla Dolphin and never touch C++ again :)

I'd be interested in hearing your ideas on how it should be implemented.

I'd be interested in seeing how you'd built subtitle file for video scenes. It started as single file to single text mapping, as I was dealing with thousands of short sound files.

FYI, I have experimented with using time instead of file offset. It works... as long as you can maintain emulation speed near 100%, which makes it virtually worthless. Syncing to emulation speed is just hell, so my best idea was to add logging mode which would dump timestamp+offset into logs and provide some simple tool that would build dummy subtitle file(s) full of offset-stamps ready to be filled with TL.

Project as a whole has been put on hold due to health issues, but I am slowly coming back online.

**Comment by PTwr** [comment]
Hi, Dolphin subtitle system is a dirty hack abusing its notification system. Its just reusing whatever system was in place to display system messages. As such there are multiple shortcomings, like crappy pixel-font that looks horribly when text is larger.
I think I remember bits of code where feade-out behavior could be changed.

As you probably noticed, subtitle format is rather crappy and has to struggle against limitations of both Wii and rest of Dolphin.
It is all just an experimental implementation using dirty hack, and movie-subs are even more experimental. To put it bluntly, I had no idea what I was doing, what I needed, and how to glue it all together, and then kept crashing against all the weird quirks of emulation.

It is unfinished, and highly experimental, but I plan to get back to it "soon". Rough roadmap would be:

    start from scratch on newest Dolphin sourcode
    port simplified text display thingie
    fix all the memory leaks in JSON parsing...
    improve file offset logger
    make MR with some nice examples of usage to get it into vanilla Dolphin and never touch C++ again :)

I'd be interested in hearing your ideas on how it should be implemented.

I'd be interested in seeing how you'd built subtitle file for video scenes. It started as single file to single text mapping, as I was dealing with thousands of short sound files.

FYI, I have experimented with using time instead of file offset. It works... as long as you can maintain emulation speed near 100%, which makes it virtually worthless. Syncing to emulation speed is just hell, so my best idea was to add logging mode which would dump timestamp+offset into logs and provide some simple tool that would build dummy subtitle file(s) full of offset-stamps ready to be filled with TL.

Project as a whole has been put on hold due to health issues, but I am slowly coming back online.

**Comment by radianthero** [comment]
Thanks for the reply, glad you're feeling better. 

The only suggestions I could think of would be

Remove the fade out, my use case is for an fmv, and it makes it hard to follow.

Even if not possible, editing the font to be easier to read, or have a drop shadow would be nice as well for subs. 

Making timing easier to write out, maybe a nitpick but I found the Ms over s timing to be annoying to type to, so an option for neither or could be good. 

Also your json tool is amazing for translating a bunch of text easily, its a lot better vs manual editing with asm.

**Comment by PTwr** [comment]
Oh, I'm glad that weirdness of that json works out for you too! It was purpose built to help keeping subs for 6 thousand voice files somewhat organized :D

Fade out should be easy change from what I remember of related Dolphin code.
Better font and control over background is well withing possibilities of in-game UI framework Dolphin uses, its just not implemented/exposed enough for subtitle system to make use currently. It would require more digging in depths of Dolphin codebase.

**Comment by radianthero** [comment]
Not sure how newGbaXL didn't, but he was able to use bin strings to convert all the original jp to a text file to easily replace it with ascii text. It took way less time to edit things, so that's been a big help. 

Sadly I can't find someone to translate the last 2 chapters, but that's a different issue lol. Either way, thank you for your contributions.

**Comment by PTwr** [comment]
Hah, BinString tool is an abomination that should have never been made... but it works :D
And its CLI is half-broken, gotta rewrite it too. I want to give it some UI too, so regexp tweaking would give visible results in real time.

Turns out that, aside of angry internet people, nothing stops you from loading unprintable characters into string then doing RegExp on ASCII codes. Eg. null delimited strings are just `[\0x01-\0xFF]+\0x00` with Single Line flag, its quite easy and you can easily do it in any language with good String and RegExp libraries.

Hard part are text encodings. When looking for actual text you gotta recode bytes to target text encoding, execute regexp, and then guestimate where in byte stream text is located. Its not an exact science with Shift-JIS encoding, as its non-reversible variable-width encoding. Recoding Shift-JIS back and forth garbles bytes a bit but length remains usually the same, so it mostly works.

edit:

In case of Lupin compiled game logic its surprisingly easy and surprisingly hard at same time :)

For example, in disc1/common/sfil1.bin you can find string entries in Notepad++ without any recoding with
```regex
(?<stringopcode>3)(?<length>[\x01-\xFF])(?<text>[\x01-\xFF]+)(?<nullterminator>\x00)
```
(Its quite loose regex so it returns all the strings, and bunch of false-positives, not just textbox content. Actual screen text is preceded by [s0/1/2/3...] commands so final regex was a lot more complex)
![image](https://github.com/PTwr/0079tl/assets/20748035/4fe5490d-4dc3-4125-9a71-f205ea7d64b5)

Its of course an unreadable mess, that can't be edited nor translated, so you gotta recode
![image](https://github.com/PTwr/0079tl/assets/20748035/732103ec-2487-4cb5-bea0-9337d65a2eee)

From this point however, we are dealing with unicode/utf charspace, so you gotta mind the regex flavors.

C#/.NET has native support for unicode, so you can just plop
```regex
(?<stringopcode>3)(?<length>[\x01-\xFF])(?<text>[\u0001-\uFFFF]+)(?<nullterminator>\x00)
```
For Notepad++ you have to search for 16bit words instead
```regex
(?<stringopcode>3)(?<length>[\x01-\xFF])(?<text>[\x{0001}-\x{FFFF}]+)(?<nullterminator>\x00)
```

Notepad2 regex engine is so dumb you can actually go for `3[\x01-\xFF][\x01-\xFF]+\x00` :D

Afterwards however goes hard part. You can't edit recoded file. Converting ascii->jis->ascii is will screw up all the binary data. Black magic is required to find string position in original byte stream.
The trick is to recode everything preceeding the string back to ascii to guestimate position in original stream, and to calculate original byte length the same way. It is surprisingly easy in C#/.NET as it has both robust text encoding tools and fixed-width 16bit char.

![image](https://github.com/PTwr/0079tl/assets/20748035/7ea3a83e-3a2b-44bb-a5a6-7fd5dfbe2ebf)

Long story short, it is not a technique one can use without some arcane encoding known-how which I somehow acquired by accident o_O

Yeah, you might take that as confirmation I am working on new version :)
